### PR TITLE
DRY: use one type class to get MIME information

### PR DESCRIPTION
**Pull request by romanofski**
We needed two separate functions in order to gather information from
MIMEMessage or WireEntity in order to render them as a list of
attachments. Both are used in separate views: the former when viewing
mails stored in notmuch the latter when constructing new mails.

This patch adds a unified interface so we can use one function to render
both.

Fixes: https://github.com/purebred-mua/purebred/issues/264

### CODE REVIEW

**Comment by romanofski** [comment]
Huh... randomly failing tests? I think I messed up showing only the filename which causes the random failures. Some paths are long enough to match the substring, some are not and then we have a failure ...

File: src/Purebred/Mail.hs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import UI.Utils (takeFileName)
+import Data.MIME
+
+-- | Provide metadata used for render the message using a unified
+-- interface
+class MIMEData a where
+  contentType :: a -> T.Text
+  filename :: CharsetLookup -> a -> T.Text
+  attachmentType :: a -> T.Text
```
Comment (frasertweedale) [inline src/Purebred/Mail.hs]: I'm really confused as to why this abstraction is necessary.  All of these only read from the headers, right?
`MIMEMessage` and `WireEntity` are both type synonyms for `Message something something'`, so you should be able to use *exactly* the same functions to read these data from either type.

Comment (romanofski) [reply src/Purebred/Mail.hs]: hrm.. odd.. I'll have another look. Cheers!

File: src/UI/Mail/Main.hs [hunk 2]
```diff
@@ -67,25 +68,22 @@ renderAttachmentsList :: AppState -> Widget Name
...
       widget = hBox
         [ padLeftRight 1 attachmentType
         , padRight Max (txt pFilename)
         , txt pType
         ]
   in withAttr listItemAttr widget
 
-isAttachment' :: WireEntity -> Bool
-isAttachment' = has (headers . contentDisposition . dispositionType . filtered (== Attachment))
+isAttachment' :: Headers -> Bool
```
Comment (romanofski) [inline src/UI/Mail/Main.hs]: This is currently implemented in purebred-email as:
```Haskell
isAttachment :: MIMEMessage -> Bool
```
but I think we should change it to accept `Headers` as the first argument. Do you concur?

Comment (frasertweedale) [reply src/UI/Mail/Main.hs]: In my 'stuff' branch it is already generalised to `isAttachment :: HasHeaders a => a -> Bool`.

`Headers` and `Message s a` have instances of `HasHeaders`.

So... soon :)

**Comment by romanofski** [comment]
Oh @frasertweedale I think this is really good to go. I'll merge if you don't mind.

**Comment by frasertweedale** [comment]
@romanofski go ahead.

**Pull request merged**